_source/_posts/2022-01-13-spring-cloud-contract.md

```diff
@@ -0,0 +1,893 @@
+
+{% include toc.md %}
+
+Imagine a simple microservice with a producer and a consumer. When writing tests in the consumer project, you have to write mocks or stubs that model the behavior of the producer project, while conversely, when you write tests in the producer project, you have to write mocks or stubs that model the behavior of the consumer project. As such, multiple sets of related, redundant code have to be maintained in parallel in disconnected projects.
+
+For every change made to the producer project, the developer needs to go to the consumer project -- a totally separate code base -- and make sure that the integration tests there reflect the new behavior producer. Keeping this up across a complex system with many different services, at best, requires a lot of discipline, and at worst, is unlikely. It's very easy for integration tests to give a false sense of security and not actually reflect the reality of the service being tested.
+
+What is needed is a way to ensure that the mocked (or stubbed) behavior of the test objects is connected to the actual behavior of the objects being modeled. Ideally, these mocked objects would be defined in the project they represent.
+
+Spring Cloud Contract does exactly this. It provides a way of specifying specific behavior for a service -- the "contract" -- and testing for compliance with this contract **within that project itself** while allowing other projects to load these contracts as stubs that can be mocked for integration testing. 
+
+{% img blog/spring-cloud-contract/spring-cloud-contract-diagram.png alt:"Spring Cloud Contract Diagram" width:"700" %}{: .center-image }
+
+A brief aside on "mocks" and "stubs". "Mock" gets used in ways online that is somewhat generic, meaning any fake object used for testing, and this can get confusing when differentiating "mocks" from "stubs". However, specifically, a "mock" is an object that tests for behavior by registering expected method calls before a test run while a "stub" is a testable version of the object with callable methods that return pre-set values. Thus, a mock checks to see if the object being tested makes an expected sequence of calls to the object being mocked, and throws an error if the behavior deviates (that is, makes any unexpected calls), while a stub does not do any testing itself, per se, but instead will return canned responses to pre-determined methods to allow tests to run (of course, it can throw an error if unexpected calls are made as well, so stubs do constrain behavior in that sense as well). 
```

Comment:
```suggestion
A brief aside on "mocks" and "stubs". "Mock" gets used in ways online that is somewhat generic, meaning any fake object used for testing, and this can get confusing when differentiating "mocks" from "stubs". However, specifically, a "mock" is an object that tests for behavior by registering expected method calls before a test run. In contrast, a "stub" is a testable version of the object with callable methods that return pre-set values. Thus, a mock checks to see if the object being tested makes an expected sequence of calls to the object being mocked, and throws an error if the behavior deviates (that is, makes any unexpected calls). A stub does not do any testing itself, per se, but instead will return canned responses to pre-determined methods to allow tests to run. (Of course, it can throw an error if unexpected calls are made as well, so stubs do constrain behavior in that sense as well). 
```
Breaking up several complex, multi-clause sentences to increase readability, especially for readers less familiar with the technologies and terminologies in question.